### DISCUSSION TITLE
Help with auto-unreference

### DISCUSSION THREAD

**Discussion by Ethan-Arrowood** [Ethan-Arrowood/undici-fetch#41] [Q&A] [answered]
I'm trying to use the new unref feature added in undici v4-rc.1

I've implemented a similar unref test that works as expected: https://github.com/Ethan-Arrowood/undici-fetch/blob/new-api/test/fetch/unref.js

Output:
```
TAP version 13
# Subtest: client unreferences self when idle
    1..1
undici-fetch
0
    ok 1 - should be equal
ok 1 - client unreferences self when idle # time=185.093ms

1..1
# time=191.285ms
```

But then when I try to implement a basic GET test it hangs for the 4 second timeout: https://github.com/Ethan-Arrowood/undici-fetch/blob/new-api/test/fetch/get.js

Output:
```
TAP version 13
# Subtest: GET request
    1..2
(node:28090) DeprecationWarning: tearDown() is deprecated, use teardown() instead
(Use `node --trace-deprecation ...` to show where the warning was created)
    ok 1 - should be equal
    ok 2 - should be equal
ok 1 - GET request # time=29.708ms

1..1
# time=4041.972ms
```

Any ideas @ronag @mcollina ? (FWIW this is all on the new-api branch if you want to dig into the code)

**Comment by mcollina** [comment]
what's the problem? I didn't get it. The test pass and there is a tap deprecation.

**Reply by ronag** [reply]
You can’t have the server and client in the same process. The server will keep the ref I think.

**Reply by mcollina** [reply]
I think we are keeping a socket ref()ed.

**Reply by Ethan-Arrowood** [reply]
the problem is that the second test suite finishes in 29ms but then doesn't exit for 4seconds (client is not unreferencing).

@ronag so the server is holding on to the socket ref even after the server shuts down? Hmm I'll have to dig into node http server and see if theres a way to unreference that

**Reply by ronag** [reply]
The client still has an open connection and close won’t kill it. It shouldn’t kill it or the test doesn’t actually test what you want. I think the only way is to have them as separate event loops.

**Comment by mcollina** [comment, marked as answer]
If you want to avoid the problem, you need to add a `Connection: close` header to tell `undici` to close the connection.

**Reply by ronag** [reply]
But that kind of kills the purpose of the test.

**Reply by mcollina** [reply]
How about:

```js
const tap = require('tap')
const http = require('http')

const { fetch } = require('../../src/fetch')

tap.test('GET request', t => {
  t.plan(2)
  const wanted = 'undici-fetch'

  const server = http.createServer((req, res) => {
    t.equal(req.method, 'GET')
    res.write(wanted)
    res.end()
    res.socket.unref() // <-- the socket definitely keeps the process open
  })

  server.unref() <-- this might or might not be needed

  t.teardown(server.close.bind(server))

  server.listen(0, () => {
    fetch(`http://localhost:${server.address().port}/`)
      .then(response => response.text())
      .then(found => {
        t.equal(found, wanted)
      })
  })
})

**Reply by Ethan-Arrowood** [reply]
Adding `res.socket.unref()` worked perfectly! Thank you @mcollina
